0005-ordered-quantification.md

```diff
@@ -0,0 +1,661 @@
+is excellent for future optimizations, but makes it harder to reason about runtime performance.
+
+It can be argued that if any compiled Dafny program uses existing features that depend on domain enumeration,
+such as `exists` expressions or `:|` statements, its runtime behavior already depends heavily on how this domain
+searching is implemented. The compiler only enforces that such domains are finite and enumerable, but an inefficient
+choice can lead to the search for a matching value taking orders of magnitude more time than expected.
+Therefore we should already be documenting this aspect of Dafny semantics and implementation better anyway, potentially
+including communicating this through the IDE.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+The primary rationale for this design is to leverage the fact that Dafny already heavily relies on expressing
+quantification directly in the language, to the point that it already has a similar `forall` statement
+expressed using the same quantification syntax. Allowing quantification to also
+declare sequential bindings provides a lot of expressive power for a small language change,
+and is just as composable with existing language features.
+
+The most obvious alternative is to only provide the simpler `foreach x <- C` syntax, with only a single bound variable,
+no additional filtering, and with `C` directly providing the enumeration and not just the ordering of values.
+This is certainly a well-formed and sound construct, but far less powerful than the proposed version, especially
+when sequence comprehensions are excluded. Offering `foreach` loops along with sequence comprehensions means we can 
+recommend the latter where possible, with the former as a fallback that is inevitably more complicated to verify.
+
+An earlier version of this proposal (https://github.com/dafny-lang/rfcs/pull/9) described an alternative
+where the ordering and multiplicity of bindings was specified by clauses such as `x in C`
+in a quantifier domain's range expression. 
+This had the advantage of not requiring any new syntax, 
+but also broke expectations of referential transparency of boolean expressions in these ordered contexts.
+
+# Prior art
+[prior-art]: #prior-art
+
+Sequence comprehensions bear a strong resemblance to list comprehensions in Python, which also include similar abilities to 
+bind multiple variables and filter the enumerated values: `[(x, y) for x in [1,2,3] for y in [3,1,4] if x != y]`.
+Haskell and Cryptol also feature similar constructs.
+
+Although the syntax `for[each] x in c` is more common, there is precedent for using the symbol `<-` instead in other languages,
+including Scala, Cryptol, and Haskell. Using something other than `in` is recommended to avoid confusion with the `in` boolean operator,
+but I don't have a strong opinion about which symbol or keyword we use.
```

Comment:
I like `<-`. (If we wanted a keyword version instead of a operator-looking version, we could use `from`. This would read well in both `x from S` and `x from 0 to 10`.)

Reply:
Btw, just to make it clear, what I said about `from` is just an observation. I'm _not_ suggesting we use the `from` keyword. In all other places in Dafny, any kind of syntax for bindings uses a symbol (in particular, `:=`, `:-`, or `:|`), and so `<-` fits better with the language.